### ISSUE TITLE
Allow @ngInject to be written as a string literal

### ISSUE THREAD

**Issue by diwu1989** [olov/ng-annotate#124]
I use coffeescript and the comment style ngInject doesn't work with coffeescript classes because the comment gets moved around too much by the compiler.

I'd like to support something like this:
```coffee
class FooController
  '@ngInject'
  constructor: ($scope, Filter) ->
    return
```
which compiles to
```js
var FooController;

FooController = (function() {
  '@ngInject';
  function FooController($scope, Filter) {
    return;
  }

  return FooController;

})();
```

I have a plugin that does this:
```js
'use strict'

var ctx;
module.exports = {
    init: function(_ctx) {
        ctx = _ctx;
    },
    match: function(node) {
      // sentinel is the string literal '@ngInject'
      if (node.type !== 'Literal' || node.value !== '@ngInject')
        return

      // found the right string literal, check that the parent is expression statement
      var expression = node.$parent;
      if (!expression || expression.type !== 'ExpressionStatement')
        return;

      // check that the grand parent is block statement
      var block = expression.$parent;
      if (!block || block.type !== 'BlockStatement')
        return;

      // find the function declaration right after the expression statement
      var body = block.body;
      if (!body)
        return;

      var declaration = body[body.indexOf(expression) + 1];
      if (!declaration || !ctx.isFunctionDeclarationWithArgs(declaration))
          return;

      // annotate the function that comes right after '@ngInject'
      ctx.addModuleContextIndependentSuspect(declaration, ctx);
      return [declaration];
    }
};
```

can you enhance ngInject to support this?

**Comment by olov** [comment]
Thanks for filing the issue. It's awesome that you were able to prototype this using a plugin!

I like the idea of using string literals as an alternative to `/*@ngInject*/` and `ngInject()`. What I would like is for such a solution to be usable for CoffeScript as well as ES6 classes (issue #64), if possible.

I came up with an alternative idea and would like to hear your thoughts on it from a CoffeScript-perspective: The `"ngInject"` string literal can be used the same way as `"use strict"`, i.e. at the beginning of a function. It will be possible to combine with other directives so `function myctrl($scope) { "use strict"; "ngInject"; ... }` will work. For ES6 classes this means that one will be able do to:

```javascript
class Foo {
    constructor($scope) { "ngInject"; ... }
}
```

compiling with Traceur into:
```javascript
$traceurRuntime.ModuleStore.getAnonymousModule(function() {
  "use strict";
  var Foo = function Foo($scope) {
    "ngInject";
    ...
  };
  ($traceurRuntime.createClass)(Foo, {}, {});
  return {};
});
```

Your CS example would become:
```coffescript
class FooController
  constructor: ($scope, Filter) ->
    'ngInject'
    return
```

and compile to
```javascript
var FooController;

FooController = (function() {
  function FooController($scope, Filter) {
    'ngInject';
    return;
  }

  return FooController;

})();
```

Thoughts?

**Comment by qraynaud** [comment]
@olov, I find this proposal great. :+1:

**Comment by diwu1989** [comment]
Having the literal inside the function looks good to me.

Thanks

**Comment by demisx** [comment]
:+1:

**Comment by olov** [comment]
Alright, I'll implement it and we'll see what happens

**Comment by mendelk** [comment]
:+1:

**Comment by olov** [comment]
"ngInject" directive prologue released in 0.15

**Comment by andreasgrimm** [comment]
sorry for being late on this, string literal sounds awesome!

any tip on how to test it right now?
right now `build.sh` throws me this error: `./build.sh: 6: ./build.sh: Syntax error: "(" unexpected`

**Comment by omsmith** [comment]
@andreasgrimm It's released in _0.15_ so you can just grab it from npm

**Comment by andreasgrimm** [comment]
awesome. thanks for pointing out. As I didn't see a release on github here I thought it isn't released yet. I'm going to test this right away ...

**Comment by olov** [comment]
@andreasgrimm I didn't `git push --tags` for 0.15 but that's done now so the releases should show up in github. As for the build error - thanks for pointing it out. I'll update the shebang.

**Comment by andreasgrimm** [comment]
awesome, thanks!

**Comment by andreasgrimm** [comment]
I encountered a problem described in https://github.com/olov/ng-annotate/issues/128

**Comment by andreasgrimm** [comment]
testing the "ngInject" directive prologue for an ES6 class constructor, I encountered the following problem:

without annotating, traceur produces this output:

```javascript
execute: function() {
  MyService = (function() {
    var MyService = function MyService($log, myRepository) {
      'ngInject';
      // ...
    };
    return ($traceurRuntime.createClass)(MyService, {}, {});
  }());
  $__export('default', MyService);
}
```
... and it works just fine.

running it through ng-annotate, this is the result:
```javascript
execute: function() {
  MyService = (function() {
    var MyService = ['$log', 'myRepository', function MyService($log, myRepository) {
      'ngInject';
      // ...
    }];
    return ($traceurRuntime.createClass)(MyService, {}, {});
  }());
  $__export('default', MyService);
}
```
... giving `$traceurRuntime.createClass` an array instead of a (constructor) function.

using `// @ngInject` right before the constructor won't work for known reasons (comments get stripped by traceur transpiler).

If I got the problem right .. how else could we make ng-annotate not using the array style but instead the `Foo.$inject = [ ...` style?
using it directly in untranspiled code like `constructor.$inject = ...` of course throws `Unexpected token .`, but it'd work on the transpiled `MyService` function I guess.

Any thoughts? Did anyone get this to work for ES6 class constructors and can tell how it is done?
